### ISSUE TITLE
Supertokens NestJs GraphQl - Cannot set headers after they are sent to the client

### ISSUE THREAD

**Issue by KonstntinM** [supertokens/supertokens-node#331]
I am trying to implement an AuthGuard for session verification in my NestJs project, that exposes a GraphQl API. I have used the [Code Snippet of the Passwordless Recipe](https://supertokens.com/docs/passwordless/nestjs/guide#8-add-session-verification-guard) for the AuthGuard and adjusted it to work with the GraphQl execution context.

This works fine for me and I am able to use all the supertokens functions with this solution, but as soon as the access token expires the program throws the following error. ```step(generator["throw"](value)); TypeError: next is not a function```. When returning false instead of throwing an error at the first if clause, the program throws ```Cannot set headers after they are sent to the client```.

I suspect that because of my use of the GraphQl Execution context the library is not able to set the cookies to refresh the access token before the response is sent. Does anyone have experience with this use case or has any idea how to solve this issue?

I am using
* Node v. 15.5.0
* Nestjs Graphql v. ^10.0.11
* supertokens-node v. ^9.2.0

This is the code of my AuthGuard

```ts
import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common';
import { GqlExecutionContext } from '@nestjs/graphql';
import { Error as STError } from 'supertokens-node';

import { verifySession } from 'supertokens-node/recipe/session/framework/express';

@Injectable()
export class AuthGuard implements CanActivate {
  async canActivate(context: ExecutionContext): Promise<boolean> {
    const gqlCtx = GqlExecutionContext.create(context).getContext();
    let err = undefined;

    const resp = gqlCtx.req.res;
    await verifySession()(gqlCtx.req, resp, (res) => {
      err = res;
    });

    if (resp.headersSent) {
      throw new STError({
        message: 'RESPONSE_SENT',
        type: 'RESPONSE_SENT',
      });
    }

    if (err) {
      return false;
    }

    return true;
  }
}
```

**Comment by rishabhpoddar** [comment]
hey @KonstntinM thanks! We will have a look ASAP

**Comment by porcellus** [comment]
Hi. Sorry for the (very) late response, but other priorities got in the way.

I've managed to find a way to solve this by adding an apollo-server plugin. It's fairly hacky, but it worked for me locally:
```ts

    GraphQLModule.forRoot<ApolloDriverConfig>({
      //  ...
      plugins: [
        {
          async requestDidStart(service) {
            return {
              async responseForOperation(requestContext) {
                let err = undefined;
                let resp = {
                  headers: new Headers(),
                  status: 200,
                  data: {},
                };

                let requestForST = {
                  headers: Object.fromEntries(requestContext.request.http.headers),
                  method: requestContext.request.http.method,
                  url: requestContext.request.http.url,
                } as any

                const respForST = {
                  getHeader(key) {
                    return resp.headers[key];
                  },
                  getHeaders() {
                    return resp.headers;
                  },
                  setHeader(key, value) {
                    resp.headers.set(key, value);
                    return this;
                  },
                  status(status) {
                    resp.status = status;
                    return this;
                  },
                  json(content) {
                    resp.data = content;
                    return this;
                  },
                } as any;
                try {
                  await verifySession({
                    sessionRequired: false,
                  })(requestForST,
                    respForST,
                    (innerErr) => {
                      err= innerErr;
                    },
                  );
                } catch (ex) {
                  err = ex;
                }

                if (err) {
                  throw err;
                }
                if (resp.status !== 200) {
                  return {
                    http: resp,
                    errors: [resp.data as any],
                  };
                }
                requestContext.context.session = requestForST.session;
                return null;
              },
            };
          },
        },
      ],
    }),
```

This replaces the auth guard usage in your resolvers, and you can just use the `context.session` to access the session if the user has one.
I'll add finding and providing a better pre-built solution to my backlog.

**Comment by porcellus** [comment]
I'm closing this for now, feel free to re-open it if you need any further help :)

**Comment by leonard-henriquez** [comment]
The solution offered by @porcellus worked for me, thank you !

But it doesn't feel right to need this snippet.
However i'm not clear if the problem comes from Supertokens or Apollo Server.
Since I found the same error message in different context, I opened an issue on Apollo Server but it might be interesting to tell them about your solution @porcellus.
How did you come with this solution? 

Package versions:
- node 16.17.0
- supertokens-node: 12.0.5
- @nestjs/graphql: 10.1.5
